src/linker/Linker.Dataflow/FlowAnnotations.cs

```diff
@@ -670,29 +672,76 @@ public FieldAnnotation (FieldDefinition field, DynamicallyAccessedMemberTypes an
 
 		internal partial MethodReturnValue GetMethodReturnValue (MethodProxy method)
-			=> GetMethodReturnValue (method, _context.Annotations.FlowAnnotations.GetReturnParameterAnnotation (method.Method));
+			=> GetMethodReturnValue (method, GetReturnParameterAnnotation (method.Method));
 
 		internal partial GenericParameterValue GetGenericParameterValue (GenericParameterProxy genericParameter)
-			=> new GenericParameterValue (genericParameter.GenericParameter, _context.Annotations.FlowAnnotations.GetGenericParameterAnnotation (genericParameter.GenericParameter));
+			=> new GenericParameterValue (genericParameter.GenericParameter, GetGenericParameterAnnotation (genericParameter.GenericParameter));
 
 #pragma warning disable CA1822 // Mark members as static - keep this an instance method for consistency with the others
 		internal partial MethodThisParameterValue GetMethodThisParameterValue (MethodProxy method, DynamicallyAccessedMemberTypes dynamicallyAccessedMemberTypes)
 			=> new MethodThisParameterValue (method.Method, dynamicallyAccessedMemberTypes);
 #pragma warning restore CA1822
 
 		internal partial MethodThisParameterValue GetMethodThisParameterValue (MethodProxy method)
-			=> GetMethodThisParameterValue (method, _context.Annotations.FlowAnnotations.GetParameterAnnotation (method.Method, 0));
+			=> GetMethodThisParameterValue (method, GetParameterAnnotation (method.Method, 0));
 
 		internal partial MethodParameterValue GetMethodParameterValue (MethodProxy method, int parameterIndex, DynamicallyAccessedMemberTypes dynamicallyAccessedMemberTypes)
 			=> new (method.Method.Parameters[parameterIndex].ParameterType.ResolveToTypeDefinition (_context), method.Method, parameterIndex, dynamicallyAccessedMemberTypes);
 
 		internal partial MethodParameterValue GetMethodParameterValue (MethodProxy method, int parameterIndex)
-			=> GetMethodParameterValue (method, parameterIndex, _context.Annotations.FlowAnnotations.GetParameterAnnotation (method.Method, parameterIndex + (method.IsStatic () ? 0 : 1)));
+			=> GetMethodParameterValue (method, parameterIndex, GetParameterAnnotation (method.Method, parameterIndex + (method.IsStatic () ? 0 : 1)));
+
+		// Linker-specific dataflow value creation. Eventually more of these should be shared.
+
+		internal SingleValue GetFieldValue (FieldDefinition field)
+		{
+			switch (field.Name) {
+			case "EmptyTypes" when field.DeclaringType.IsTypeOf (WellKnownType.System_Type): {
+					return ArrayValue.Create (0, field.DeclaringType);
+				}
+			case "Empty" when field.DeclaringType.IsTypeOf (WellKnownType.System_String): {
+					return new KnownStringValue (string.Empty);
+				}
+
+			default: {
+					DynamicallyAccessedMemberTypes memberTypes = GetFieldAnnotation (field);
+					return new FieldValue (field.FieldType.ResolveToTypeDefinition (_context), field, memberTypes);
+				}
+			}
+		}
+
+		internal SingleValue GetTypeValueNodeFromGenericArgument (TypeReference genericArgument)
```

Comment:
Nit: Can you please rename this - there are `ValueNode` things anymore - so probably just `GetTypeValueFromGenericArgument`.